share/doc/homebrew/Maintainer-Guidelines.md

```diff
@@ -6,20 +6,118 @@ definitely not a beginner’s guide.
 
 Maybe you were looking for the [Formula Cookbook](Formula-Cookbook.md)?
 
+----
+
+Quick links:
+  [General Guidelines](#general-guidelines) •
+  [Core and Core PRs](#core-and-core-prs) •
+  [Formulae and Formula PRs](#formulae-and-formula-prs)
+
+## General Guidelines
+
 ## Quick Checklist
+1.  [Ensure you have set your username and email address
+    properly](https://help.github.com/articles/setting-your-email-in-git/).
+2.  Sign off cherry-picks if you amended them
+    ([GitX-dev](https://github.com/rowanj/gitx) can do this, otherwise there is
+    a command line flag for it).
+3.  If the commit fixes a bug, use “Fixes \#104” syntax to close the bug
+    report and link to the commit.
+
+### Add Comments
+It may be enough to refer to an issue ticket, but make sure changes that would
+make sense to you, if you came to them unaware of the surrounding issues. Many
+times on other projects I’ve seen code removed because the new guy didn’t know
+why it was there. Regressions suck.
+
+### Merging, Rebasing, Cherry-Picking
+Merging is mainly useful when new work is being done. Please use `brew pull`
+(or `rebase`/`cherry-pick` contributions) rather than fill Homebrew's Git
+history up with noisy merge commits.
+
+Don’t `rebase` until you finally `push`. Once `master` is pushed, you can’t
+`rebase`: **you’re a maintainer now!**
+
+Cherry-picking changes the date of the commit, which kind of sucks.
+
+Don’t `merge` unclean branches. So if someone is still learning `git`
+their branch is filled with nonsensical merges, then `rebase` and squash
+the commits. Our main branch history should be useful to other people,
+not confusing.
+
+### Don’t Allow Bloated Diffs
```

Comment:
This is shown as an addition, but only because I had to rearrange a few sections to better separate the formula-related and general guidelines. Hence, this is the original title of that section.

“Disallow” sounds wrong to me while “Don't Allow” sounds alright. Maybe a native speaker can comment on this? In any case, I'm not opposed to changing it if it improves things.